core/log/convergence.cpp

```diff
@@ -86,25 +114,29 @@ void Convergence<ValueType>::on_criterion_check_completed(
                                            dim<2>{1, residual->get_size()[1]});
                     dense_r->compute_norm2(this->residual_norm_);
                 });
+        } else if (dynamic_cast<const solver::SolverBase*>(solver) &&
+                   b != nullptr && x != nullptr) {
```

Comment:
I think we will also need the IterationLogger for getting iteration only.
Convergence will compute the residual norm (with additional residual computation in some cases)

Reply:
I don't think this is relevant to this PR. The convergence logger should compute the residual norm, that is the point of it. I added this case, because sometimes the residual is not available (e.g. MG or first check in IR). Now the logger should be able to always compute the residual.

Reply:
IterationLogger should be in another pr, I mean.
If the user just needs the iteration count, the convergence requires more operation than what the user needs.